lib/foreman_remote_execution_core/polling_script_runner.rb

```diff
@@ -0,0 +1,138 @@
+      publish_data(err, 'stderr') unless err.empty?
+      if result
+        exitcode = result[1] || 0
+        publish_exit_status(exitcode.to_i)
+      end
+      destroy_session
+    end
+
+    def external_event(event)
+      publish_data(event.data['output'], 'stdout')
+      publish_exit_status(event.data['exit_code'].to_i)
+      super
+    end
+
+    def close
+      super
+      SmartProxyDynflowCore::OtpManager.drop_otp(@uuid, @otp) if @otp
+    end
+
+    def prepare_retrieval
+      # The script always outputs at least one line
+      # First line of the output either has to begin with
+      # "RUNNING" or "DONE $EXITCODE"
+      # The following lines are treated as regular output
+      base = File.dirname(@output_path)
+      posfile = File.join(base, 'position')
+      tmpfile = File.join(base, 'tmp')
+      script = <<-SCRIPT.gsub(/^ +\| /, '')
+      | #!/bin/sh
+      | pid=$(cat "#{@pid_path}")
+      | if ! pgrep --help 2>/dev/null >/dev/null; then
+      |   echo DONE 1
+      |   echo "pgrep is required" >&2
+      |   exit 1
+      | fi
+      | if pgrep -P "$pid" >/dev/null 2>/dev/null; then
+      |   echo RUNNING
+      | else
+      |   echo "DONE $(cat "#{@exit_code_path}" 2>/dev/null)"
+      | fi
+      | [ -f "#{@output_path}" ] || exit 0
+      | [ -f "#{posfile}" ] || echo 1 > "#{posfile}"
+      | position=$(cat "#{posfile}")
+      | tail --bytes "+${position}" "#{@output_path}" > "#{tmpfile}"
+      | bytes=$(cat "#{tmpfile}" | wc --bytes)
+      | expr "${position}" + "${bytes}" > "#{posfile}"
+      | cat "#{tmpfile}"
+      SCRIPT
+      @logger.debug("copying script:\n#{script.lines.map { |line| "    | #{line}" }.join}")
+      cp_script_to_remote(script, 'retrieve')
+      @prepared = true
+    end
+
+    def callback_scriptlet(callback_script_path = nil)
+      if @otp
+        callback_script_path = cp_script_to_remote(callback_script, 'callback') if callback_script_path.nil?
+        "#{su_prefix}#{callback_script_path}"
+      else
+        ':' # Shell synonym for "do nothing"
+      end
+    end
+
+    def callback_script
+      <<-SCRIPT.gsub(/^ +\| /, '')
+      | #!/bin/sh
+      | exit_code=$(cat "#{@exit_code_path}")
+      | url="#{@callback_host}/dynflow/tasks/#{@task_id}/done"
+      | output="$(cat "#{@output_path}" | base64 --wrap 0)"
+      | json="{\\\"step_id\\\": #{@step_id}, \\\"output\\\": \\\"$output\\\", \\\"exit_code\\\": $exit_code }"
+      | if which curl >/dev/null; then
+      |   curl -X POST -d "$json" -u "#{@task_id}:#{@otp}" "$url"
+      | else
+      |   echo 'curl is required' >&2
+      |   exit 1
+      | fi
+      SCRIPT
+    end
+
+    private
+
+    def destroy_session
+      if @session
+        @logger.debug("Closing session with #{@ssh_user}@#{@host}")
+        @session.close
+        @session = nil
+      end
+    end
+
+    def try_obtain_otp!
+      @otp = SmartProxyDynflowCore::OtpManager.get_otp(@uuid) if defined? SmartProxyDynflowCore
```

Comment:
What should happen when the `SmartProxyDynflowCore` is not present? Shouldn't the `otp` logic go to `foreman_tasks_core` instead?